File: Level 4/트리 트리오 중간값/solution.py
from collections import deque


def build_tree(edges):
    tree = {}

    for start, end in edges:
        if not start in tree:
            tree[start] = []
        if not end in tree:
            tree[end] = []
        tree[start].append(end)
        tree[end].append(start)

    return tree


def bfs(tree, start):
    result = []
    visited = set([start])
    queue = deque()
    queue.append((start, 0))

    while queue:
        cur, distance = queue.popleft()
        result.append((cur, distance))

        for node in tree[cur]:
            if not node in visited:
                queue.append((node, distance + 1))
                visited.add(node)

    return result


def solution(n, edges):
    tree = build_tree(edges)

    leaf = bfs(tree, 1)
    father = bfs(tree, leaf[-1][0])

    return father[-1][1] if leaf[-1][1] == leaf[-2][1] else father[-2][1]

File: Level 4/트리 트리오 중간값/solution.js
function solution(n, edges) {
  const tree = buildTree(edges);

  const leaf = bfs(tree, 1);
  const father = bfs(tree, leaf.at(-1)[0]);

  return leaf.at(-1)[1] === leaf.at(-2)[1]
    ? father.at(-1)[1]
    : father.at(-2)[1];
}

function buildTree(edges) {
  const tree = {};

  edges.forEach(([start, end]) => {
    if (!tree[start]) tree[start] = [];
    if (!tree[end]) tree[end] = [];

    tree[start].push(end);
    tree[end].push(start);
  });

  return tree;
}

function bfs(tree, start) {
  const result = [];
  const visited = {};
  const queue = new Queue();

  queue.push([start, 0]);
  visited[start] = true;

  while (queue.size()) {
    const [cur, distance] = queue.popleft();
    result.push([cur, distance]);

    tree[cur].forEach((node) => {
      if (!visited[node]) {
        queue.push([node, distance + 1]);
        visited[node] = true;
      }
    });
  }

  return result;
}

class Queue {
  constructor() {
    this.storage = new Map();
    this.front = 0;
    this.rear = 0;
  }

  size() {
    return this.storage.size;
  }

  push(value) {
    if (!this.storage.size) {
      this.storage.set(0, value);
    } else {
      this.rear += 1;
      this.storage.set(this.rear, value);
    }
  }

  popleft() {
    const item = this.storage.get(this.front);

    if (this.storage.size === 1) {
      this.storage.clear();

      this.front = 0;
      this.rear = 0;
    } else {
      this.storage.delete(this.front);
      this.front += 1;
    }

    return item;
  }
}
